File: C#/C#-Advanced-SoftUni/StacksAndQueues/BalancedParentheses.cs
using System;
using System.Collections.Generic;

namespace BalancedParentheses
{

    //има грешка
    class BalancedParentheses
    {
        static void Main(string[] args)
        {
            Stack<char> leftSide = new Stack<char>(Console.ReadLine());
            int originaLength = leftSide.Count;
            Stack<char> rightSide = new Stack<char>();


            if (originaLength % 2 == 0)
            {
                for (int i = 0; i < originaLength / 2; i++)
                {
                    char parentheses = leftSide.Pop();
                    rightSide.Push(parentheses);
                }

                bool isBalanced = true;

                int length = leftSide.Count; //Math.Min(leftSide.Count, rightSide.Count);

                for (int i = 0; i < length; i++)
                {
                    char left = leftSide.Pop();
                    char right = rightSide.Pop();

                    if (left == '[' && right == ']')
                    {

                    }
                    else if (left == '{' && right == '}')
                    {

                    }
                    else if (left == '(' && right == ')')
                    {

                    }
                    else
                    {
                        isBalanced = false;
                        break;
                    }
                }

                if (isBalanced && leftSide.Count == rightSide.Count)
                {
                    Console.WriteLine("YES");
                }
                else
                {
                    Console.WriteLine("NO");
                }
            }
            else
            {
                Console.WriteLine("NO");
            }
        }
    }
}

File: Python/Advanced/stacks_queues/balanced_parentheses.py
from collections import deque

parentheses = input()
parentheses_stack = []
parentheses_list = []
parentheses_list[:0] = parentheses
is_balanced = True

opened_parentheses = 0
parentheses_queue = deque(parentheses_list)

while parentheses_queue:
    char = parentheses_queue.popleft()

    if char == '(' or char == '[' or char == '{':
        parentheses_stack.append(char)
    elif char == ')' and parentheses_stack:
        if parentheses_stack[-1] == '(':
            parentheses_stack.pop()
        else:
            is_balanced = False
            break
    elif char == '}' and parentheses_stack:
        if parentheses_stack[-1] == '{':
            parentheses_stack.pop()
        else:
            is_balanced = False
            break
    elif char == ']' and parentheses_stack:
        if parentheses_stack[-1] == '[':
            parentheses_stack.pop()
        else:
            is_balanced = False
            break
    else:
        is_balanced = False
        break

if parentheses_stack or not is_balanced:
    print('NO')
else:
    print('YES')
